torch/csrc/autograd/python_variable_indexing.cpp

```diff
@@ -63,197 +64,154 @@ static int64_t count_specified_dimensions(PyObject* index) {
+      throw python_error();
+    }
+    tensor_index.set_slice({start, stop, step});
+  } else if (obj == Py_Ellipsis) {
+    tensor_index.set_ellipsis();
+  } else if (obj == Py_None) {
+    tensor_index.set_none();
+  } else if (PyBool_Check(obj)) {
+    tensor_index.set_boolean(obj == Py_True);
+  } else if (THPVariable_Check(obj)) {
+    tensor_index.set_tensor(THPVariable_Unpack(obj));
+  } else if (PySequence_Check(obj)) {
+    // TODO: Naughty naughty get out of jail free
+    // (Fixing this means I have to fix the call chain though :/)
+    tensor_index.set_tensor(sequenceToVariable(legacyExtractDispatchKey(self), obj));
   } else {
-    return at::empty({0}, self.options().dtype(kLong));
+    tensor_index.set_integer(THPUtils_unpackLong(convertToPythonInt(obj)));
   }
 }
 
-static Variable applySlicing(const Variable& self, PyObject* index, variable_list& outIndices) {
+static inline Variable applySlicing(const Variable& self, PyObject* index, variable_list& outIndices, bool is_tracing) {
   int64_t size = PyTuple_GET_SIZE(index); // NOLINT(cppcoreguidelines-pro-type-cstyle-cast)
   int64_t dim = 0;
   int64_t specified_dims = count_specified_dimensions(index);
 
-  auto handle_var = [&](const Variable& var) {
-    // TODO: check scalarType
-    outIndices.resize(dim + 1);
-    outIndices[dim] = var;
-    dim++;
-  };
-
   if (specified_dims > self.dim()) {
     throw IndexError("too many indices for tensor of dimension %d", (int)self.dim());
   }
 
   Variable result = self;
   for (int64_t i = 0; i < size; i++) {
     PyObject* obj = PyTuple_GET_ITEM(index, i); // NOLINT(cppcoreguidelines-pro-type-cstyle-cast)
-    if (THPUtils_checkLong(obj)) {
-      result = applySelect(result, dim, obj, i);
-    } else if (PySlice_Check(obj)) {
-      result = applySlice(result, dim, obj);
-      dim++;
-    } else if (obj == Py_Ellipsis) {
-      dim += self.dim() - specified_dims;
-    } else if (obj == Py_None) {
-      result = result.unsqueeze(dim);
-      dim++;
-    } else if (PyBool_Check(obj)) {
-      result = result.unsqueeze(dim);
-      handle_var(boolToIndexingTensor(result, obj == Py_True)); // NOLINT(cppcoreguidelines-pro-type-cstyle-cast)
-    } else if (THPVariable_Check(obj)) {
-      auto& var = THPVariable_Unpack(obj);
-      auto scalar_type = var.scalar_type();
-      if (var.dim() == 0 && at::isIntegralType(scalar_type, /*includeBool=*/true)) {
-        if (scalar_type != at::kByte && scalar_type != at::kBool) {
-          result = applySelect(result, dim, obj, i);
-        } else {
-          result = result.unsqueeze(dim);
-          if(scalar_type == at::kBool) {
-            handle_var(boolToIndexingTensor(result, var.item<bool>() != 0));
-          } else {
-            handle_var(boolToIndexingTensor(result, var.item<uint8_t>() != 0));
-          }
+
+    // Handle JIT tracing
+    if (is_tracing) {
+      if (THPUtils_checkLong(obj) && THPVariable_Check(obj)) {
+        recordSelectTrace(THPVariable_Unpack(obj));
```

Comment:
Is it possible that the tracing code should have been moved into `indexToTensorIndex`? It's hard for me to tell as it is kind of different